cmd/oaas/handler_build_test.go

```diff
@@ -207,3 +211,110 @@ func TestHandleIncludedSourcesBadTypes(t *testing.T) {
+# simulate failure
+echo "err on stdout"
+>&2 echo "err on stderr"
+exit 23
+`)
+	defer restore()
+
+	buf := makeTestPost(t, `{"exports": ["tree"], "environments": ["MY=env"]}`, `{"fake": "manifest"}`)
+	rsp, err := http.Post(endpoint, "application/x-tar", buf)
+	assert.NoError(t, err)
+	defer ioutil.ReadAll(rsp.Body)
+
+	assert.Equal(t, rsp.StatusCode, http.StatusCreated)
+	reader := bufio.NewReader(rsp.Body)
+	content, err := ioutil.ReadAll(reader)
+	assert.NoError(t, err)
+	expectedContent := fmt.Sprintf(`starting %[1]s/build build
+err on stdout
+err on stderr
+cannot run osbuild: exit status 23`, buildDir)
+	assert.Equal(t, expectedContent, string(content))
+
+	// check that the result is an error and we get the log
+	endpoint = baseURL + "api/v1/result/image/disk.img"
+	rsp, err = http.Get(endpoint)
+	assert.NoError(t, err)
+	assert.Equal(t, http.StatusBadRequest, rsp.StatusCode)
+	reader = bufio.NewReader(rsp.Body)
+	content, err = ioutil.ReadAll(reader)
+	assert.NoError(t, err)
+	assert.Equal(t, "build failed\n"+expectedContent, string(content))
+}
+
+func TestBuildStreamsOutput(t *testing.T) {
+	baseURL, baseBuildDir, _ := runTestServer(t)
+	endpoint := baseURL + "api/v1/build"
+
+	restore := main.MockOsbuildBinary(t, fmt.Sprintf(`#!/bin/sh -e
+for i in $(seq 5); do
+   echo "line-$i"
+   sleep 0.2
+done
+
+# simulate output
+mkdir -p %[1]s/build/output/image
+echo "fake-build-result" > %[1]s/build/output/image/disk.img
+`, baseBuildDir))
+	defer restore()
+
+	buf := makeTestPost(t, `{"exports": ["tree"], "environments": ["MY=env"]}`, `{"fake": "manifest"}`)
+	rsp, err := http.Post(endpoint, "application/x-tar", buf)
+	assert.NoError(t, err)
+	defer ioutil.ReadAll(rsp.Body)
+
+	assert.Equal(t, rsp.StatusCode, http.StatusCreated)
+	reader := bufio.NewReader(rsp.Body)
+	line, err := reader.ReadString('\n')
+	assert.NoError(t, err)
+	headerLine := fmt.Sprintf("starting %s/build build\n", baseBuildDir)
+	assert.Equal(t, headerLine, line)
+
+	// This is not ideal, as it's timing dependend, ideally the script
+	// would signal somehow that it just echoed a line and the test
+	// code then expects to recieve it within e.g. 100ms.
+	start := time.Now()
+	for i := 1; i < 6; i++ {
+		line, err := reader.ReadString('\n')
+		assert.NoError(t, err)
+		assert.Equal(t, fmt.Sprintf("line-%v\n", i), line)
+		timeSinceStart := time.Now().Sub(start)
+		// we expeced a line every 200ms and add a grace
+		// period of 100ms for slow systems (should be plenty
+		// but super busy GH runners can take a long time to
+		// do anything)
+		expectedTimeSinceStart := time.Duration(i)*200*time.Millisecond + (100 * time.Millisecond)
+		assert.True(t, timeSinceStart < expectedTimeSinceStart, fmt.Sprintf("time since start %v bigger than expected %v", timeSinceStart, expectedTimeSinceStart))
+	}
+}
+
+func TestBuildErrorHandlingTar(t *testing.T) {
+	restore := main.MockOsbuildBinary(t, `#!/bin/sh
+
+# not creating an output dir, this will lead to errors from the "tar"
```

Comment:
Not ideal, better would be to mock tar directly but I feel a bit lazy today :/

Reply:
intimately familiar with that feeling